src/goto-instrument/replace_runtime_checks.cpp

```diff
@@ -0,0 +1,209 @@
+/*******************************************************************\
+
+Module: Replace Runtime Checks
+
+Author: diffblue
+
+\*******************************************************************/
+
+/// \file
+/// Replace Runtime Checks
+
+#include <util/cprover_prefix.h>
+#include <util/prefix.h>
+#include <util/string_constant.h>
+
+#include <langapi/language.h>
+#include <langapi/mode.h>
+
+#include "replace_runtime_checks.h"
+
+/// A simple pass that replace each call to a function with name:
+/// __CPROVER_X_`check-name` with a user-specified sequence of instructions. The
+/// language X can also by specified by the user via
+/// `--runtime-checks-language`. The function is expected to take one argument
+/// (the condition to be checked) and this condition is either asserted or
+/// assumed (or both). The `check-name` is then included in the assertion
+/// description (via source-location).
+struct runtime_checks_replacert
+{
+public:
+  explicit runtime_checks_replacert(
+    goto_modelt &goto_model,
+    const namespacet &ns,
+    const std::string &language_id)
+    : goto_model(goto_model),
+      ns(ns),
+      language_id(language_id),
+      common_prefix(std::string{CPROVER_PREFIX} + language_id)
+  {
+  }
+
+  void replace_with(const std::string &replace_option);
+
+private:
+  goto_modelt &goto_model;
+  const namespacet &ns;
+  const std::string language_id;
+  const std::string common_prefix;
+  irep_idt mode;
+
+  enum class replace_typet
+  {
+    assert_only,
+    assume_only,
+    assert_then_assume
+  };
+
+  enum class runtime_check_typet
+  {
+    overflow,
+    range,
+    division_by_zero
+  };
```

Comment:
Do we need to enumerate these here?  IIRC there are a number of others and it would be a little bit annoying to have to do parallel maintenance between this and gnat2goto.